crates/ruff_linter/src/noqa.rs

```diff
@@ -418,182 +300,408 @@ impl<'a> FileNoqaDirectives<'a> {
+
+                    // Whitespace could be a delimiter or the end of the `noqa`.
+                    // If it's the end, then non-ascii whitespace is allowed
+                    c if c.is_whitespace() => false,
+
+                    // e.g. #noqa:F401F842
+                    //                ^
+                    // Push a warning and update the `missing_delimiter`
+                    // state but don't consume the character since it's
+                    // part of the next code.
+                    c if c.is_ascii_uppercase() => {
+                        self.warnings
+                            .push(LexicalWarning::MissingDelimiter(TextRange::empty(
+                                self.offset + self.current(),
+                            )));
+                        true
+                    }
+                    // Start of a new comment
+                    // e.g. #noqa: F401#A comment
+                    //                 ^
+                    '#' => false,
+                    _ => return Err(LexicalError::InvalidCodeSuffix),
+                };
+            }
+            Some(_) => {
+                // The first time we hit an evidently invalid code,
+                // it's probably a trailing comment. So stop lexing,
+                // but don't push an error.
+                // Ex)
+                // # noqa: F401 we import this for a reason
+                //              ^----- stop lexing but no error
+                self.cursor.skip_bytes(self.cursor.as_str().len());
+            }
+            None => {}
+        }
+        Ok(())
     }
 
-    /// Lex the "ruff" prefix of a `noqa` directive.
-    #[inline]
-    fn lex_ruff(line: &str) -> Option<&str> {
-        line.strip_prefix("ruff")
+    /// Push current token to stack of [`Code`] objects
+    fn push_code(&mut self) {
+        self.codes.push(Code {
+            code: &self.line[self.token_range()],
+            range: self.token_range().add(self.offset),
+        });
     }
 
-    /// Lex a `noqa` directive with case-insensitive matching.
+    /// Current token range relative to offset
     #[inline]
-    fn lex_noqa(line: &str) -> Option<&str> {
-        match line.as_bytes() {
-            [b'n' | b'N', b'o' | b'O', b'q' | b'Q', b'a' | b'A', ..] => Some(&line["noqa".len()..]),
-            _ => None,
-        }
+    fn token_range(&self) -> TextRange {
+        let end = self.current();
+        let len = self.cursor.token_len();
+
+        TextRange::at(end - len, len)
     }
 
-    /// Lex a code delimiter, which can either be a comma or whitespace.
+    /// Retrieves the current position of the cursor within the line.
     #[inline]
-    fn lex_delimiter(line: &str) -> Option<&str> {
-        let mut chars = line.chars();
-        if let Some(c) = chars.next() {
-            if c == ',' || c.is_whitespace() {
-                Some(chars.as_str())
-            } else {
-                None
+    fn current(&self) -> TextSize {
```

Comment:
Nit: I'd rename this to `position` or `offset`. It's unclear if `current` refers to the current offset or current character or something else.